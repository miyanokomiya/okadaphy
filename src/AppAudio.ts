// https://umipla.com/
// @ts-ignore
import slashSoundURL1 from '../assets/se_sword_hit9.mp3'
// @ts-ignore
import slashSoundURL2 from '../assets/se_sword_hit2.mp3'

export class AppAudio {
  private slashSound1: HTMLAudioElement
  private slashSound2: HTMLAudioElement

  constructor() {
    this.slashSound1 = new Audio(slashSoundURL1)
    this.slashSound1.load()
    this.slashSound2 = new Audio(slashSoundURL2)
    this.slashSound2.load()
  }

  public playSlash(seed: number): void {
    this.slashSound1.pause()
    this.slashSound1.currentTime = 0
    this.slashSound2.pause()
    this.slashSound2.currentTime = 0

    if (seed < 0.5) {
      this.slashSound1.play()
    } else {
      this.slashSound2.play()
    }
  }
}
