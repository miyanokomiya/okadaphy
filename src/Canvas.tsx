import * as React from 'react'
import PropTypes from 'prop-types'
import App from './app'
import { IOptions } from '../types'

type Props = IOptions & {
  width?: number
  height?: number
  count?: number
}

const Canvas: React.FC<Props> = props => {
  const canvasRef = React.useRef(null)

  React.useEffect(() => {
    const app = new App({ canvas: canvasRef.current as any })
    app.importFromString({
      fillStyle: props.fillStyle,
      strokeStyle: props.strokeStyle,
      text: props.text,
    })
    app.setGravity(props.gravityX, props.gravityY)
    app.run()
    return () => app.dispose()
  }, [props])

  return <canvas ref={canvasRef} width={props.width} height={props.height} />
}

Canvas.propTypes = {
  width: PropTypes.number,
  height: PropTypes.number,
  text: PropTypes.string.isRequired,
  fillStyle: PropTypes.string.isRequired,
  strokeStyle: PropTypes.string.isRequired,
  gravityX: PropTypes.number.isRequired,
  gravityY: PropTypes.number.isRequired,
  count: PropTypes.number,
}
Canvas.defaultProps = {
  width: 300,
  height: 300,
  count: 0,
}

export default Canvas
