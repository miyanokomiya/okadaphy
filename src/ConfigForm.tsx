import * as React from 'react'
import PropTypes from 'prop-types'
import { IOptions } from '../types'

type Props = IOptions & {
  onSubmit: (options: IOptions) => void
}

const OptionForm: React.FC<Props> = props => {
  const [draftText, setDraftText] = React.useState(props.text)
  const [draftFillStyle, setDraftFillStyle] = React.useState(props.fillStyle)
  const [draftStrokeStyle, setDraftStrokeStyle] = React.useState(props.strokeStyle)
  const [draftGravityX, setDraftGravityX] = React.useState(props.gravityX)
  const [draftGravityY, setDraftGravityY] = React.useState(props.gravityY)

  const onInputDraftText = React.useCallback((e: React.FormEvent<HTMLTextAreaElement>) => {
    setDraftText(e.currentTarget.value)
  }, [])
  const onInputDraftFillStyle = React.useCallback((e: React.FormEvent<HTMLInputElement>) => {
    setDraftFillStyle(e.currentTarget.value)
  }, [])
  const onInputDraftStrokeStyle = React.useCallback((e: React.FormEvent<HTMLInputElement>) => {
    setDraftStrokeStyle(e.currentTarget.value)
  }, [])
  const onInputDraftGravityX = React.useCallback((e: React.FormEvent<HTMLInputElement>) => {
    setDraftGravityX(parseFloat(e.currentTarget.value))
  }, [])
  const onInputDraftGravityY = React.useCallback((e: React.FormEvent<HTMLInputElement>) => {
    setDraftGravityY(parseFloat(e.currentTarget.value))
  }, [])

  const onSubmit = React.useCallback(
    (e: React.FormEvent<HTMLFormElement>) => {
      e.preventDefault()
      props.onSubmit({
        text: draftText,
        fillStyle: draftFillStyle,
        strokeStyle: draftStrokeStyle,
        gravityX: draftGravityX,
        gravityY: draftGravityY,
      })
    },
    [props, draftText, draftFillStyle, draftStrokeStyle, draftGravityX, draftGravityY],
  )

  return (
    <form onSubmit={onSubmit}>
      <textarea cols={30} rows={3} value={draftText} onChange={onInputDraftText} />
      <button>Run</button>
      <br />
      <br />
      面:
      <input type="color" value={draftFillStyle} onChange={onInputDraftFillStyle} />
      <br />
      線:
      <input type="color" value={draftStrokeStyle} onChange={onInputDraftStrokeStyle} />
      <br />
      重力x:
      <input
        min="-0.05"
        max="0.05"
        step="0.0001"
        type="range"
        value={draftGravityX}
        onChange={onInputDraftGravityX}
      />
      <br />
      重力y:
      <input
        min="-0.05"
        max="0.05"
        step="0.0001"
        type="range"
        value={draftGravityY}
        onChange={onInputDraftGravityY}
      />
    </form>
  )
}

OptionForm.propTypes = {
  text: PropTypes.string.isRequired,
  fillStyle: PropTypes.string.isRequired,
  strokeStyle: PropTypes.string.isRequired,
  gravityX: PropTypes.number.isRequired,
  gravityY: PropTypes.number.isRequired,
  onSubmit: PropTypes.func.isRequired,
}
OptionForm.defaultProps = {}

export default OptionForm
