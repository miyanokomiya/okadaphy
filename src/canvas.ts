import type { IVec2 } from 'okageo'

export function getCursorPoint(e: MouseEvent | TouchEvent): IVec2 {
  const rect = (e.target as HTMLCanvasElement).getBoundingClientRect()
  const positionX = rect.left + window.pageXOffset
  const positionY = rect.top + window.pageYOffset

  if (e instanceof MouseEvent) {
    return {
      x: e.pageX - positionX,
      y: e.pageY - positionY,
    }
  } else {
    const touch = e.touches[0]
    return {
      x: touch.pageX - positionX,
      y: touch.pageY - positionY,
    }
  }
}
