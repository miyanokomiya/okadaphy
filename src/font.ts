import { Font, load } from 'opentype.js'
import okageo, { IVec2, ISvgPath, ISvgStyle } from 'okageo'

let _font: Font | null
const fontURL = 'https://fonts.gstatic.com/ea/notosansjapanese/v6/NotoSansJP-Medium.otf'

export function loadFont(): Promise<Font> {
  return new Promise((resolve, reject) => {
    if (_font) resolve(_font)

    load(fontURL, (err, font) => {
      if (!font || err) {
        return reject(new Error('Font could not be loaded: ' + err))
      }
      return resolve(font)
    })
  })
}

export async function parseFont(text: string, style: ISvgStyle): Promise<ISvgPath[]> {
  const font = await loadFont()
  const lines = text.split(/\n|\r\n/)
  const size = 72
  let pathList: IVec2[][] = []
  lines.forEach((line, i) => {
    pathList = pathList.concat(
      okageo.svg
        .parseOpenPath(font.getPath(line, 0, size * 1.1 * i, size))
        .map((info) => okageo.geo.omitSamePoint(info.d)),
    )
  })

  return okageo.geo.getIncludedPolygonGroups(pathList).map((group) => {
    const [d, ...included] = group
    return { d, included, style }
  })
}
