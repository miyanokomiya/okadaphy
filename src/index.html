<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>岡田を切る</title>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="./index.tsx"></script>
  </body>
</html>
