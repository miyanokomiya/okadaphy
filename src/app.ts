import { Body as MBody, Engine, Events, IEventCollision, Runner, World } from 'matter-js'
import { ISvgPath, IVec2 } from 'okageo'
import * as geo from 'okageo/src/geo'
import * as svg from 'okageo/src/svg'
import * as opentype from 'opentype.js'
import { IBodyShape, ISlash } from '../types/index'
import { drawFrame, FRAME_DEPTH, getFrameBodies } from './frame'
import { createShape, mergeShape, splitShape } from './shape'

// matterがdecompを使うが、parcelのせいかimportがうまくいかない
(window as any).decomp = require('poly-decomp')

svg.configs.bezierSplitSize = 3
svg.configs.ellipseSplitSize = 3

export default class App {
  private canvas: HTMLCanvasElement
  private clearEventListener: () => void
  private ctx: CanvasRenderingContext2D
  private cursorDownPoint: IVec2 | null
  private cursorMovePoint: IVec2 | null
  private engine: Engine
  private runner: Runner
  private running: boolean
  private shapeList: IBodyShape[]
  private slashList: ISlash[]
  private font: any

  constructor (args: { canvas: HTMLCanvasElement }) {
    this.canvas = args.canvas
    this.clearEventListener = () => { return }
    this.ctx = args.canvas.getContext('2d') as CanvasRenderingContext2D
    this.cursorDownPoint = null
    this.cursorMovePoint = null
    this.engine = Engine.create()
    this.engine.world.gravity.scale = 0
    this.runner = Runner.create({})
    this.running = false
    this.shapeList = []
    this.slashList = []

    // 壁生成
    World.add(this.engine.world, getFrameBodies(this.canvas.width, this.canvas.height))
    this.draw()
    this.stop()
    this.initEventListener()
  }

  public importFromDefault () {
    const pathInfoList = [{
      d: [{ x: 0, y: 0 }, { x: 10, y: 0 }, { x: 10, y: 10 }, { x: 0, y: 10 }],
      style: {
        fill: true,
        fillGlobalAlpha: 1,
        fillStyle: 'gray',
        lineCap: 'butt',
        lineDash: [],
        lineJoin: 'bevel',
        lineWidth: 1,
        stroke: true,
        strokeGlobalAlpha: 1,
        strokeStyle: 'yellow'
      }
    }]
    const margin = FRAME_DEPTH * 14
    const inRectList = svg.fitRect(
      pathInfoList,
      margin,
      margin,
      this.canvas.width - margin * 2,
      this.canvas.height - margin * 2
    )
    inRectList.forEach((info) => {
      const s = createShape(info)
      if (s) {
        this.addShape(s)
      }
    })
    this.shapeList.forEach((shape) => {
      MBody.setAngularVelocity(shape.body, 0.01)
    })
    this.draw()
  }

  public importFromSVG (svgStr: string) {
    const pathInfoList = svg.parseSvgGraphicsStr(svgStr)
    const margin = FRAME_DEPTH
    const inRectList = svg.fitRect(
      pathInfoList,
      margin,
      margin,
      this.canvas.width - margin * 2,
      this.canvas.height - margin * 2
    )
    inRectList.forEach((info) => {
      const s = createShape(info)
      if (s) {
        this.addShape(s)
      }
    })
    this.draw()
  }

  public importFromString (str: string) {
    this.loadFont().then(() => {
      const lines = str.split(/\n|\r\n/)
      const size = 72
      let pathList: IVec2[][] = []
      lines.forEach((line, i) => {
        const top = size * 1.1 * i
        const fontPath = this.font.getPath(line, 0, top, size)
        pathList = pathList.concat(
          svg.parseOpenPath(fontPath).map((info) => geo.omitSamePoint(info.d)))
      })

      const groups = geo.getIncludedPolygonGroups(pathList)
      const style = {
        ...svg.createStyle(),
        fill: true,
        fillStyle: 'black', stroke: false
      }
      const pathInfoList: ISvgPath[] = groups.map((group) => {
        const [d, ...included] = group
        return { d, included, style }
      })
      const margin = FRAME_DEPTH + 30
      svg.fitRect(
        pathInfoList,
        margin,
        margin,
        this.canvas.width - margin * 2,
        this.canvas.height - margin * 2
      ).forEach((info) => {
        const s = createShape(info)
        if (s) {
          this.addShape(s)
        }
      })
      this.draw()
    })
  }

  public run () {
    Runner.start(this.runner, this.engine)
    this.running = true
  }

  public stop () {
    Runner.stop(this.runner)
    this.running = false
  }

  public clear () {
    Engine.clear(this.engine)
    this.cursorDownPoint = null
    this.shapeList.concat().forEach((s) => this.removeShape(s))
    this.shapeList = []
    this.slashList = []
  }

  public dispose () {
    this.stop()
    Engine.clear(this.engine)
    this.running = false
    this.clearEventListener()
  }

  public isRunning () {
    return this.running
  }

  private loadFont (): Promise<undefined> {
    return new Promise((resolve, reject) => {
      if (this.font) return resolve()

      opentype.load('https://fonts.gstatic.com/ea/notosansjapanese/v6/NotoSansJP-Medium.otf', (err, font) => {
        if (!font || err) {
          return reject(new Error('Font could not be loaded: ' + err))
        }
        this.font = font
        return resolve()
      })
    })
  }

  private draw () {
    if (!this.ctx) return
    this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height)
    drawFrame(this.ctx)

    this.shapeList.forEach((shape) => {
      this.ctx.save()
      this.ctx.translate(shape.body.position.x, shape.body.position.y)
      this.ctx.rotate(shape.body.angle)
      svg.draw(this.ctx, {
        d: shape.vertices,
        included: shape.included,
        style: shape.style
      })
      this.ctx.restore()
    })

    this.ctx.strokeStyle = 'yellow'
    this.ctx.lineWidth = 2
    this.slashList.forEach((slash) => {
      this.ctx.beginPath()
      this.ctx.moveTo(slash.line[0].x, slash.line[0].y)
      this.ctx.lineTo(slash.line[1].x, slash.line[1].y)
      this.ctx.stroke()
    })
  }

  private initEventListener () {
    const onCursorDown = (e: MouseEvent | TouchEvent) => this.onCursorDown(e)
    const onCursorUp = (e: MouseEvent | TouchEvent) => this.onCursorUp(e)
    const onCursorMove = (e: MouseEvent | TouchEvent) => this.onCursorMove(e)
    this.canvas.addEventListener('mousedown', onCursorDown, true)
    this.canvas.addEventListener('touchstart', onCursorDown, true)
    this.canvas.addEventListener('mouseup', onCursorUp, true)
    this.canvas.addEventListener('touchend', onCursorUp, true)
    this.canvas.addEventListener('mousemove', onCursorMove, true)
    this.canvas.addEventListener('touchmove', onCursorMove, true)
    this.clearEventListener = () => {
      this.canvas.removeEventListener('mousedown', onCursorDown, true)
      this.canvas.removeEventListener('touchstart', onCursorDown, true)
      this.canvas.removeEventListener('mouseup', onCursorUp, true)
      this.canvas.removeEventListener('touchend', onCursorUp, true)
      this.canvas.removeEventListener('mousemove', onCursorMove, true)
      this.canvas.removeEventListener('touchmove', onCursorMove, true)
    }

    Events.on(this.engine, 'afterUpdate', () => this.afterUpdate())
    // Events.on(this.engine, 'collisionActive', (e) => this.collisionActive(e))
  }

  private addShape (shape: IBodyShape): void {
    World.add(this.engine.world, [shape.body])
    this.shapeList.push(shape)
  }

  private removeShape (shape: IBodyShape): void {
    World.remove(this.engine.world, shape.body)
    this.shapeList.splice(this.shapeList.indexOf(shape), 1)
  }

  private collisionActive (e: IEventCollision<Engine>) {
    e.pairs
      .filter((pair) => pair.timeUpdated - pair.timeCreated > 3000)
      .forEach((pair) => {
        if (Math.random() < 0.990) return

        const shapeA = this.findShape(pair.bodyA.id)
        const shapeB = this.findShape(pair.bodyB.id)
        if (!shapeA || !shapeB) return

        const mergedShape: IBodyShape | null = mergeShape(shapeA, shapeB)
        if (!mergedShape) return

        this.removeShape(shapeA)
        this.removeShape(shapeB)
        this.addShape(mergedShape)
      })
  }

  private findShape (id: number): IBodyShape | null {
    let shape: IBodyShape | null = null
    this.shapeList.some((s) => {
      // partsとして抱えるbodyも検索対象とする
      s.body.parts.some((part) => {
        if (part.id === id) shape = s
        return !!shape
      })
      return !!shape
    })
    return shape
  }

  private afterUpdate () {
    // this.pullCenter()
    // this.pullEachOther()
    this.slashList.forEach((slash) => slash.time++)
    this.slashList = this.slashList.filter((slash) => slash.time < 60)
    this.draw()
  }

  private pullEachOther () {
    this.shapeList.forEach((base) => {
      this.shapeList.forEach((shape) => {
        if (base.body.id === shape.body.id) return

        const vec = geo.sub(base.body.position, shape.body.position)
        const d = geo.getNorm(vec)
        if (d < 1) return
        const force = geo.multi(vec, 0.00001 * base.body.mass / Math.pow(d, 2))
        MBody.applyForce(shape.body, shape.body.position, force)
      })
    })
  }

  private pullCenter () {
    const center = { x: this.canvas.width / 2, y: this.canvas.height / 2 }
    this.shapeList.forEach((shape) => {
      const vec = geo.sub(center, shape.body.position)
      const d = geo.getNorm(vec)
      if (d < 1) return
      const force = geo.multi(vec, 0.00001 / Math.pow(d, 2))
      MBody.applyForce(shape.body, shape.body.position, force)
    })
  }

  private onCursorDown (e: MouseEvent | TouchEvent) {
    e.preventDefault()
    this.cursorDownPoint = getCursorPoint(e)
  }

  private onCursorMove (e: MouseEvent | TouchEvent) {
    e.preventDefault()
    this.cursorMovePoint = getCursorPoint(e)
  }

  private onCursorUp (e: MouseEvent | TouchEvent) {
    e.preventDefault()
    if (!this.cursorDownPoint || !this.cursorMovePoint) return

    if (geo.isSame(this.cursorDownPoint, this.cursorMovePoint)) return

    this.slash(expandLine(this.cursorDownPoint, this.cursorMovePoint))
    this.cursorDownPoint = null
  }

  private slash (line: IVec2[]) {
    this.slashList.push({ line, time: 0 })
    this.shapeList.concat().forEach((shape) => {
      const splitedShapeList = splitShape(shape, line)
      if (!splitedShapeList) return

      // 分割前を削除して分割後を追加
      this.removeShape(shape)
      splitedShapeList.forEach((s) => this.addShape(s))
    })

    this.draw()
  }
}

function getCursorPoint (e: MouseEvent | TouchEvent): IVec2 {
  const rect = (e.target as HTMLCanvasElement).getBoundingClientRect()
  const positionX = rect.left + window.pageXOffset
  const positionY = rect.top + window.pageYOffset

  if (e instanceof TouchEvent) {
    const touch = e.touches[0]
    return {
      x : touch.pageX - positionX,
      y : touch.pageY - positionY
    }
  } else {
    return {
      x : e.pageX - positionX,
      y : e.pageY - positionY
    }
  }
}

function expandLine (a: IVec2, b: IVec2): IVec2[] {
  const v = geo.multi(geo.getUnit(geo.sub(b, a)), 4000)
  return [geo.sub(a, v), geo.add(a, v)]
}
