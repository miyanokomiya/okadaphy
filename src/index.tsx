import * as React from 'react'
import { render } from 'react-dom'
import Canvas from './Canvas'
import ConfigForm from './ConfigForm'
import { IOptions } from '../types'

const App: React.FC = () => {
  const [count, refresh] = React.useState(0)
  const [text, setText] = React.useState('岡田を\n切る技術')
  const [fillStyle, setFillStyle] = React.useState('#000')
  const [strokeStyle, setStrokeStyle] = React.useState('#fefb00')
  const [gravityX, setGravityX] = React.useState(0)
  const [gravityY, setGravityY] = React.useState(0)
  const onSubmit = React.useCallback(
    (options: IOptions) => {
      setText(options.text)
      setFillStyle(options.fillStyle)
      setStrokeStyle(options.strokeStyle)
      setGravityX(options.gravityX)
      setGravityY(options.gravityY)
      refresh(count + 1)
    },
    [count],
  )

  return (
    <div>
      <Canvas
        width={window.innerWidth * 0.8}
        height={innerHeight * 0.7}
        text={text}
        fillStyle={fillStyle}
        strokeStyle={strokeStyle}
        gravityX={gravityX}
        gravityY={gravityY}
        count={count}
      />
      <ConfigForm
        text={text}
        fillStyle={fillStyle}
        strokeStyle={strokeStyle}
        gravityX={gravityX}
        gravityY={gravityY}
        onSubmit={onSubmit}
      />
    </div>
  )
}

render(<App />, document.getElementById('root') as any)
