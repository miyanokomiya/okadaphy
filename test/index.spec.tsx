import * as React from 'react'
import renderer from 'react-test-renderer'
import { App } from '../src/index'

jest.mock('@material-ui/core/Slider', () => 'Slider')
jest.mock('@material-ui/core/TextField', () => 'TextField')
jest.mock('react-color', () => ({ CompactPicker: 'CompactPicker' }))

describe('src/components/ConfigForm', () => {
  it('snapshot', () => {
    const tree = renderer.create(<App />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
