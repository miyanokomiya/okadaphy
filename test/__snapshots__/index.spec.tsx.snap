// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/ConfigForm snapshot 1`] = `
<div>
  <div>
    <canvas
      height={537.5999999999999}
      width={1008}
    />
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
    >
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Pause
          </span>
        </button>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Step
          </span>
        </button>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "marginTop": "1rem",
      }
    }
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="MuiGrid-root MuiGrid-container"
      >
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained"
            disabled={false}
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="submit"
          >
            <span
              className="MuiButton-label"
            >
              Reset
            </span>
          </button>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-container"
        style={
          Object {
            "marginTop": "1rem",
          }
        }
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            Text
          </p>
          <TextField
            fullWidth={true}
            multiline={true}
            onChange={[Function]}
            value="岡田を
切る技術"
            variant="outlined"
          />
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
        style={
          Object {
            "marginTop": "1rem",
          }
        }
      >
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            Fill
          </p>
          <CompactPicker
            color="#000"
            onChange={[Function]}
          />
        </div>
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            Stroke
          </p>
          <CompactPicker
            color="#fefb00"
            onChange={[Function]}
          />
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
        style={
          Object {
            "marginTop": "1rem",
          }
        }
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
        >
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            Gravity X
          </p>
          <Slider
            marks={
              Array [
                Object {
                  "label": "0",
                  "value": 0,
                },
              ]
            }
            max={0.05}
            min={-0.05}
            onChange={[Function]}
            step={0.0001}
            value={0}
            valueLabelFormat={[Function]}
          />
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
        >
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            Gravity Y
          </p>
          <Slider
            marks={
              Array [
                Object {
                  "label": "0",
                  "value": 0,
                },
              ]
            }
            max={0.05}
            min={-0.05}
            onChange={[Function]}
            step={0.0001}
            value={0}
            valueLabelFormat={[Function]}
          />
        </div>
      </div>
    </form>
  </div>
  <a
    className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
    href="https://github.com/miyanokomiya/okadaphy"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    Github
  </a>
</div>
`;
